Snakemake scaffold

tkosciol opened the pull request:
addressing #5 and #22 

There are some issues highlighted in the code that need addressing.
It also needs cleaning from hard-coded paths which were included for testing.

`extra.rules` need to be added to the main Snakefile for the final version.

DO NOT MERGE, but I would appreciate comments on the code.

RNAer commented on microprot/scripts/snakemake_helpers.py:
```diff
@@ -0,0 +1,48 @@
+
+def msa_size(msa_fp):
+    msa_dir, msa_ext = os.path.splitext(os.path.abspath(msa_fp))
+    if msa_ext is not '.a3m':
```
use `==` instead of `is` for str comparison

RNAer commented on microprot/snakemake/Snakefile:
```diff
@@ -3,192 +3,222 @@
+        a3m = config['MICROPROT_MAIN']+'results/01-search_pdb/{seq}/{seq}.a3m'
     log:
-        'results/log/{seq}/search_pdb.log'
+        config['MICROPROT_MAIN']+'results/log/{seq}/search_pdb.log'
```
log file is not used anywhere

RNAer replied:
also in other rules

tkosciol commented:
@RNAer thanks so much for your quick comments!
About log files, that's something I'm aware of. I'd like to keep all of the stdout as logs. Is there a way to do it in Snakemake other than piping to logs?

RNAer commented:
I am not aware of. last time I checked it seems not

On Wed, Jun 7, 2017 at 8:37 PM, Tomasz <notifications@github.com> wrote:

> @RNAer <https://github.com/rnaer> thanks so much for your quick comments!
> About log files, that's something I'm aware of. I'd like to keep all of
> the stdout as logs. Is there a way to do it in Snakemake other than piping
> to logs?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/microprot/pull/46#issuecomment-306989386>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA9CeJb9tS_2uR8PYF8EISpU-0fdpe7rks5sB2yKgaJpZM4NlwHv>
> .
>

sjanssen2 commented on microprot/snakemake/Snakefile:
```diff
@@ -3,192 +3,276 @@
+        '%s/%s' % (SEQ_path, SEQ_file)
     output:
-        'data/{seq}.fasta'
+        config['MICROPROT_MAIN']+'data/WCG_inputs/{seq}.fasta'
```
output filepaths depend on the "name" of the sequence. What if this name contains strange characters as / or | ?

tkosciol replied:
then we have a potential problem :/
but I don't see a way around it - Snakemake bases input-output rule connections off names. Potentially, we could use made up names, but that would be even more problematic, as we would not be able to identify what we're looking at without a lookup table of some sorts.

tkosciol commented on microprot/snakemake/Snakefile:
```diff
@@ -3,192 +3,276 @@
+from shutil import copyfile
+from skbio import io
+
+sys.path.append('/projects/microprot')
```
I kept `sys.path.append` for debugging. Should disappear from the release version.
Agreed about `/` issue.
What would you suggest for input/output situation? Have explicit `MICROPROT_INPUT` and `MICROPROT_OUTPUT` definitions in the config file? But then, I want to have the directory structure kept inside results, i.e. `01-seqrch_pdb`, etc.

sjanssen2 replied:
I think it makes sense to keep the dir structure as is, but give the user the opportunity to split inputs (for which read permissions are sufficient) and all kind of results (for which also write permissions are required). Furthermore, you could have a different output dir than me, but we both operate on the same input.
Can you have a MICROPROT_OUTPUT which is set to MICROPROT_INPUT by default?

Regarding special characters: I suggest to do a replacement of the chars in the name: `name.replace(['/', '\', '|'], ['_','_','_'])`

tkosciol replied:
OK, input/output split implemented.
Need a few more minutes for `replace`. Is not as trivial as it seems.;)

tkosciol commented on microprot/snakemake/Snakefile:
```diff
@@ -3,192 +3,276 @@
     output:
-        not_ripe='results/{seq}/09-not_ripe/{seq}.msa',
-        ripe='results/{seq}/10-Rosetta/{seq}.msa'
+        config['MICROPROT_MAIN']+'results/log/{seq}/msa_ripe.log'
```
Rule's output (`msa_ripe.log`) is written regardless of what happens downstream, so we should be good here. However, it is very likely that if there is no input for one of the `search_X` rules (e.g. if the entire sequence matches PDB or CM) we will get an error. That needs further investigation.
Comments in lines 94-98 outline how to potentially deal with this problem. What do you think?

tkosciol commented on microprot/snakemake/Snakefile:
```diff
@@ -3,192 +3,276 @@
-                         -oa3m {a3m}')
-        search_X(input['seq'], input['split'], output[0], shell_f)
+        """
+        it's essentially a copy of `search_x` function, but here we're using
```
agreed, but then search_x needs to be rewritten, to take either hhblits or hhsearch - each of those programs has different commandline structure!

tkosciol commented:
I addressed most of the comments, I hope.
But here's a fun problem I've got. If there is more than 1 input file (see `/projects/microprot/benchmarking/snakemake_minimal_test/seq3.faa`) output names are correct, but contents of each file come just from the first one. I'm trying to debug this, but having another pair of eyes would help.

tkosciol commented:
@sjanssen2 it now works fine with multiple inputs. I still haven't merged in `MSA_hhblits` rule into `search_x`, but I would call it a cosmetic improvement and have more time sensitive issues to deal with.
Please, test it out on a variety of inputs and let me know your thoughts.

tkosciol commented:
@sjanssen2 @qiyunzhu @RNAer it's not pretty, but it's good for review.
Please note, this is the first version of the pipeline. So what I'm going for here it's the fact that it's working and does the job.

upcoming for future PRs:
* reorganize output structure #52 
* change `snakemake_helpers.write_db` functionality #48 
* benchmarking and results reports #57

qiyunzhu commented on microprot/scripts/process_input.py:
```diff
@@ -0,0 +1,66 @@
+import numpy as np
+import click
+import os
```
According to PEP8, built-in packages like `os` should appear before 3rd party ones like `click`.

qiyunzhu commented on microprot/scripts/process_input.py:
```diff
@@ -0,0 +1,66 @@
+
+    suffix = []
+
+    if sort is True:
```
The variable name `sort` looks so close to reserved Python keywords. Maybe do `sort_by_length`?

qiyunzhu commented on microprot/scripts/process_input.py:
```diff
@@ -0,0 +1,66 @@
+        suffix.append('_sorted')
+        s = []
+        for seq in fasta:
+            s.append(len(seq))
```
This can be`s = [len(seq) for seq in fasta]`.

qiyunzhu commented on microprot/scripts/process_input.py:
```diff
@@ -0,0 +1,66 @@
+        for i, seq in enumerate(fasta):
+            if len(seq) >= min_len and len(seq) <= max_len:
+                idx.append(i)
+        output_fasta = fasta[idx]
```
I feel this entire function is overly complicated. Why not just do `output_fasta = [seq for seq in fasta if min_len <= len(seq) <= max_len]`? It is for performance consideration?

qiyunzhu commented on microprot/scripts/process_input.py:
```diff
@@ -0,0 +1,66 @@
+
+    suffix = ''.join(suffix)
+
+    if outfile:
```
I would suggest:
```
if outfile is None:
    outfile = '%s%s.fasta' % (fp_name, suffix)
process_fasta.write_sequences(output_fasta, outfile)
```

qiyunzhu commented on microprot/scripts/snakemake_helpers.py:
```diff
@@ -0,0 +1,91 @@
+
+
+def not_empty(fname):
+    if os.path.exists(fname) and os.path.getsize(fname) > 0:
```
I suggest:
```
return os.path.exists(fname) and os.path.getsize(fname) > 0
```

qiyunzhu commented on microprot/scripts/snakemake_helpers.py:
```diff
@@ -0,0 +1,91 @@
+
+
+def msa_size(msa_fp):
+    msa_dir, msa_ext = os.path.splitext(os.path.abspath(msa_fp))
```
Why most of the functions don't have docstring?

qiyunzhu commented on microprot/scripts/snakemake_helpers.py:
```diff
@@ -0,0 +1,91 @@
+        msa_ext = '.a3m'
+    with open(''.join([msa_dir, msa_ext]), 'r') as f:
+        lines = f.readlines()
+    msa_size = len([line for line in lines if line.startswith('>')])-1
```
This may be an infficient way of using memory. How about `msa_size = sum(1 for line in lines if line.startswith('>')) - 1`.

qiyunzhu commented on microprot/scripts/snakemake_helpers.py:
```diff
@@ -0,0 +1,91 @@
+    processed_fh = open('%s/%s' % (microprot_out,
+                                   'processed_sequences.fasta'), 'a')
+    for i, SEQ in enumerate(SEQS):
+        _seq = SEQ.metadata['id']
```
I feel they can be merged in one line.

qiyunzhu commented on microprot/scripts/snakemake_helpers.py:
```diff
@@ -0,0 +1,91 @@
+                                               version,
+                                               timestamp)
+        with open('%s.index' % db_fp, 'a') as f:
+            f.write(append_idx)
```
I  used to experience issues when you don't have `\n` written in the end.

tkosciol replied:
in `append_seq` definition there is `\n`

tkosciol commented on microprot/scripts/process_input.py:
```diff
@@ -0,0 +1,66 @@
+                idx.append(i)
+        output_fasta = fasta[idx]
+
+    if min_len > 1:
```
do this in `click` instead of checking here explicitly.

tkosciol commented:
@qiyunzhu comments addressed! I hope it's ready to merge.:)

qiyunzhu commented:
@tkosciol Good job! I think it is ready to merge. Does @RNAer still request changes?

tkosciol commented:
I think I addressed all of his requests.

Tomasz Kosciolek, PhD
Postdoctoral Research Associate
Rob Knight Lab
University of California San Diego

On Jul 9, 2017, 3:26 PM -0700, Qiyun Zhu <notifications@github.com>, wrote:

@tkosciol<https://github.com/tkosciol> Good job! I think it is ready to merge. Does @RNAer<https://github.com/rnaer> still request changes?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/biocore/microprot/pull/46#issuecomment-313969651>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AGZ0Vhebwy3k6XFMmyzSel7G7N8H5Sluks5sMVOMgaJpZM4NlwHv>.

qiyunzhu commented:
@tkosciol Merged!
